hypervisor/src/arch/emulator/mod.rs

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright © 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+
+extern crate thiserror;
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum EmulatorError {
+    #[error("Unsupported instruction: {0}")]
+    UnsupportedInstruction(#[source] anyhow::Error),
+
+    #[error("Unsupported memory size: {0}")]
+    UnsupportedMemorySize(#[source] anyhow::Error),
+
+    #[error("Invalid instruction: {0}")]
+    InvalidInstruction(#[source] anyhow::Error),
+
+    #[error("Invalid register: {0}")]
+    InvalidRegister(#[source] anyhow::Error),
+
+    #[error("Invalid operand: {0}")]
+    InvalidOperand(#[source] anyhow::Error),
+
+    #[error("Wrong number of operands: {0}")]
+    WrongNumberOperands(#[source] anyhow::Error),
+}
+
+/// A RegisterEmulator implements a platform registers read and write
+/// emulation.
+pub trait RegisterEmulator: Send + Sync {
+    type Register;
+
```

Comment:
Why not do `fn read_reg<T>(&self, reg: Self::Register> -> Result<T, EmulatorError>`, Would remove the need for all the variants.